### ISSUE TITLE
`unnecessary_safety_comment` false positive in `_ = unsafe { foo_unsafe() }` expression, conflicts with `undocumented_unsafe_blocks` correct analysis

### ISSUE THREAD

**Issue by sandersaares** [rust-lang/rust-clippy#13039]
### Summary

The `unnecessary_safety_comment` lint considers a safety comment as being unnecessary if the code is of the shape:

```
_ = unsafe { foo_unsafe() }
```

The following alternative shape is correctly identified as requiring a safety comment:

```
let _ = unsafe { foo_unsafe() }
```

The `undocumented_unsafe_blocks` lint expects a safety comment in both cases, so the first case cannot pass Clippy check if both lints are enabled.

### Lint Name

unnecessary_safety_comment

### Reproducer

```rust
unsafe fn foo() -> usize {
    1234
}

#[deny(clippy::unnecessary_safety_comment)]
#[deny(clippy::undocumented_unsafe_blocks)] // For comparison
fn main() {
    // SAFETY: unnecessary_safety_comment incorrectly triggers here
    _ = unsafe { foo() };
}
```

### Version

```text
rustc 1.79.0 (c6bcfe6a60 2024-06-18) (1.79.0-ms-20240619.1+c6bcfe6a60)
binary: rustc
commit-hash: c6bcfe6a6082b23fa084d9690fee241c23ee9d29
commit-date: 2024-06-18
host: x86_64-pc-windows-msvc
release: 1.79.0
LLVM version: 18.1.7
```
